src/error.rs

```diff
@@ -4,83 +4,265 @@ use std::io;
 use std::num;
 use std::str;
 use std::string;
+use url;
+
+/// Client-side errors
+#[derive(Debug, PartialEq)]
+pub enum ClientError {
+    /// The key provided was longer than 250 bytes.
+    KeyTooLong,
+    /// The server returned an error prefixed with CLIENT_ERROR in response to a command.
+    Error(String),
+}
+
+impl fmt::Display for ClientError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ClientError::KeyTooLong => write!(f, "The provided key was too long."),
+            ClientError::Error(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+impl From<ClientError> for MemcacheError {
+    fn from(err: ClientError) -> Self {
+        MemcacheError::ClientError(err)
+    }
+}
+
+/// Server-side errors
+#[derive(Debug)]
+pub enum ServerError {
+    /// When using binary protocol, the server returned magic byte other
+    /// than 0x81 in the response packet.
+    BadMagic(u8),
+    /// The client did not expect this response from the server.
+    BadResponse(String),
+    /// The server returned an error prefixed with SERVER_ERROR in response to a command.
+    Error(String),
+}
+
+impl fmt::Display for ServerError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ServerError::BadMagic(e) => write!(f, "Expected 0x81 as magic in response header, but found: {:x}", e),
+            ServerError::BadResponse(s) => write!(f, "Unexpected: {} in response", s),
+            ServerError::Error(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+/// Command specific errors.
+#[derive(Debug, PartialEq)]
+pub enum CommandError {
+    /// The client tried to set a key which already existed in the server.
+    KeyExists,
+    /// The client tried to set a key which does not exist in the server.
+    KeyNotFound,
+    /// The value for a key was too large. The limit is usually 1MB.
+    ValueTooLarge,
+    /// Invalid arguments were passed to the command.
+    InvalidArguments,
+    /// The server requires authentication.
+    AuthenticationRequired,
+    /// When using binary protocol, the server returned an unknown response status.
+    Unknown(u16),
+    /// The client sent an invalid command to the server.
+    InvalidCommand,
+}
+
+impl MemcacheError {
+    pub(crate) fn try_from(s: String) -> Result<String, MemcacheError> {
```

Comment:
This function name and signature seems a little bit wired to me, I think it's not a function may raise an error, but a function try to parse some value.

So I think it's better to let the result to be an `Option` and function name changed to be `parse_error_string`? I'm not sure about this, what do you think?

Reply:
Again, the weirdness in signature is for convenience - to be able to use the `?`. I think `Result` is better suited here. The name is borrowed from [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html). But I cannot implement `TryFrom` for `String`.